### ISSUE TITLE
[Spec] IndicatorsView

### ISSUE THREAD

**Issue by rmarinho** [xamarin/Xamarin.Forms#6577]
# IndicatorsView

Provide a list of Indicators (shapes) to represent a number of items and a current position. This will be mainly to work with `CarouselView` #4996 

# API

## IndicatorsView
```csharp
public class IndicatorsView : View
{
	 public static readonly BindableProperty IndicatorsColorProperty = BindableProperty.Create(nameof(IndicatorsColor), typeof(Color), typeof(IndicatorsView), Color.Silver);

	public Color IndicatorsColor
	{
		get { return (Color)GetValue(IndicatorsColorProperty); }
		set { SetValue(IndicatorsColorProperty, value); }
	}

	public static readonly BindableProperty SelectedIndicatorColorProperty = BindableProperty.Create(nameof(SelectedIndicatorColor), typeof(Color), typeof(IndicatorsView), Color.Gray);

	public Color SelectedIndicatorColor
	{
		get { return (Color)GetValue(SelectedIndicatorColorProperty); }
		set { SetValue(SelectedIndicatorColorProperty, value); }
	}

	public static readonly BindableProperty IndicatorsShapeProperty = BindableProperty.Create(nameof(IndicatorsShape), typeof(IndicatorsShape), typeof(IndicatorsView), IndicatorsShape.Circle);

	public IndicatorsShape IndicatorsShape
	{
		get { return (IndicatorsShape)GetValue(IndicatorsShapeProperty); }
		set { SetValue(IndicatorsShapeProperty, value); }
	}

       public DataTemplate IndicatorTemplate
	{
		get { return (DataTemplate)GetValue(IndicatorTemplate Property); }
		set { SetValue(IndicatorTemplate Property, value); }
	}
	public static readonly BindableProperty CountProperty = BindableProperty.Create(nameof(Count), typeof(int), typeof(IndicatorsView), default(int));

	public int Count
	{
		get => (int)GetValue(CountProperty);
		set => SetValue(CountProperty, value);
	}

	 public static readonly BindableProperty MaxVisibleCountProperty = BindableProperty.Create(nameof(MaxVisibleCount), typeof(int), typeof(IndicatorsView), -1);

	public int MaxVisibleCount
	{
		get => (int)GetValue(MaxVisibleCountProperty);
		set => SetValue(MaxVisibleCountProperty, value);
	}

	public static readonly BindableProperty PositionProperty = BindableProperty.Create(nameof(Position), typeof(int), typeof(IndicatorsView), default(int), BindingMode.TwoWay, propertyChanged: OnPositionPropertyChanged);

	public int Position
	{
		get => (int)GetValue(PositionProperty);
		set => SetValue(PositionProperty, value);
	}

	public event EventHandler<PositionChangedEventArgs> PositionChanged;
}
```
### Properties

| API | Description |
| ------------- | ------------- |
| IndicatorsColor | Color of the indicators  |
| SelectedIndicatorColor | Color of the indicator that shows the selected item  |
| IndicatorShape | Shape to render on Indicator  |
| IndicatorTemplate | Template to render on Indicator  |
| Count | Number of indicators  |
| MaxVisibleCount | Number of visible indicators, defaults to -1 so Count is used  |
| Position | Current selected indicator index  |

## IndicatorShape
Specify how the Indicator should look, Circle or Square
```csharp
public enum IndicatorShape
{
    Circle,
    Square
}
```

# Scenarios
```csharp
var indicatorsView = new IndicatorsView
{
	BackgroundColor = Color.Red,
	IndicatorsTintColor = Color.Blue,
        SelectedIndicatorTintColor = Color.Yellow,
	IndicatorsShape = IndicatorsShape.Square,
	HorizontalOptions = LayoutOptions.Center,
	VerticalOptions = LayoutOptions.End,
	HeightRequest = 30,
	InputTransparent = true
};
```
```
	IndicatorsView.SetItemsSourceBy(indicatorsView, carouselView);
``` 
# Backward Compatibility
New control so no issues.

# Difficulty : [medium]

**Comment by rmarinho** [comment]
Examples of what we want to achieve 

![carousel_rei](https://user-images.githubusercontent.com/41873/51287079-6d02f580-19bb-11e9-938e-a47789266974.gif)

**Comment by andreinitescu** [comment]
There are several ways to do this, an easier way and a bit more complicated one, but much more flexible and more rewarding on the long term. 

## The flexible way 

Ideally, for maximum flexibility, this view should be **a view with selectable and templatable items**. 
Which sounds like the `SelectableItemsView`.

I wish the `SelectableItemsView` would not have multiple selection, and instead, there should be a `MultiSelectatbleItemsView`:

```
public class MultiSelectableItemsView : SelectableItemsView
{
}
```

 This way, `IndicatorsView` could really be just a specialization of the `SelectableItemsView` with default item template and item layout:

```
public IndicatorsView : SelectableItemsView
{
}
```

I'm not decided if it wouldn't be better if the `IndicatorsView` would derive from `TemplatedView` (or `View`) but it should still use a `SelectableItemsView` under the hood because it makes the implementation extremely simple by reusing existing controls).

### IndicatorsView item template

The default item template can contain a marker element, something like the `IndicatorsShape` you used:
```
<IndicatorsView ..>
   <IndicatorsView.ItemTemplate>
      <DataTemplate>
          <IndicatorShape />
      </DataTemplate>
   <IndicatorsView.ItemTemplate>
</IndicatorsView>
```
The default renderer can implement the `IndicatorsShape` by drawing the indicator (for example, using a Drawable on Android). 

Since it's a templatable item, if someone wants to, it can replace the template with an image:

```
<IndicatorsView ..>
   <IndicatorsView.ItemTemplate>
      <DataTemplate>
          <Image Source='...' />
      </DataTemplate>
   <IndicatorsView.ItemTemplate>
</IndicatorsView>
```
Additionally, there could be some standard visual states like "Selected", "Unselected" which the `SelectableItemsView` could set on the generated item-view when item is known to be selected.

### IndicatorsView items layout

Being based on the `ItemsView` gives the `IndicatorsView` a flexible way to display the items if someone needs to change this. The default layout for the `IndicatorsView` could be a `StackLayout` which has the nice `Spacing` property. 

For this, the `ItemsView` should support a `Layout<View>` as a layout. And in turn, to make this work, we should change the `Layout<View>` to implement the marker interface `IItemsLayout` interface, (which would open up really cool possibilities for creating custom controls!).

The default layout for the `IndicatorsView` could be this:
```
<IndicatorsView>
    <IndicatorsView.ItemsLayout>
             <StackLayout Spacing="10" />
    </IndicatorsView.ItemsLayout>
</IndicatorsView>
```

### ItemsIndicator item generation

Similar to the `ItemsControl` in WPF, I wish the `ItemsView` had an `Items` property in order to support the scenario of actually providing the item views. This way, the `ItemsIndicator` could actually create and sets its items based on the `Count` and `ItemTemplate` properties.

## Using ItemsIndicators without the CollectionView

The ItemsIndicators could be used in other scenarios where such indicators are needed. For example, someone has some custom way of displaying a collection of items but still needs an items indicator.

**Comment by andreinitescu** [comment]
If the item template would be based on the `DataTemplate` and the `IndicatorItem` marker element (like how I described above), the `IndicatorItem` marker should have all the properties related to the indicator:

```
public class IndicatorItem : Element
{
         public Color SelectedColor { get; set; }
         public Color UnselectedColor { get; set; }
         public IndicatorShape Shape { get; set; }
}
```

**Comment by rmarinho** [comment]
I think we should maybe just go and implement this instead https://github.com/Microsoft/microsoft-ui-xaml-specs/blob/user/savoyschuler/pagercontrol/active/PagerControl/PagerControl.md

**Comment by AndreiMisiukevich** [comment]
smth like this
```csharp
using System.Collections;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Linq;

namespace Xamarin.Forms
{
	public enum IndicatorShape
	{
		Circle,
		Square
	}

	public class IndicatorsView : StackLayout
	{
		public static readonly BindableProperty PositionProperty = BindableProperty.Create(nameof(Position), typeof(int), typeof(IndicatorsView), default(int), BindingMode.TwoWay, propertyChanged: (bindable, oldValue, newValue) =>
		{
			(bindable as IndicatorsView)?.ResetIndicatorsStyles();
		});

		public static readonly BindableProperty CountProperty = BindableProperty.Create(nameof(Count), typeof(int), typeof(IndicatorsView), default(int), propertyChanged: (bindable, oldValue, newValue) =>
		{
			(bindable as IndicatorsView)?.ResetIndicatorsCount((int)oldValue, (int)newValue);
		});

		public static readonly BindableProperty MaxVisibleCountProperty = BindableProperty.Create(nameof(MaxVisibleCount), typeof(int), typeof(IndicatorsView), int.MaxValue, propertyChanged: (bindable, oldValue, newValue) =>
		{
			(bindable as IndicatorsView)?.ResetIndicatorsStyles();
		});

		public static readonly BindableProperty IndicatorTemplateProperty = BindableProperty.Create(nameof(IndicatorTemplate), typeof(DataTemplate), typeof(IndicatorsView), default(DataTemplate), propertyChanged: (bindable, oldValue, newValue) =>
		{
			(bindable as IndicatorsView)?.ResetIndicatorsStyles();
		});

		public static readonly BindableProperty IndicatorShapeProperty = BindableProperty.Create(nameof(IndicatorShape), typeof(IndicatorShape), typeof(IndicatorsView), default(IndicatorShape), propertyChanged: (bindable, oldValue, newValue) =>
		{
			(bindable as IndicatorsView)?.ResetIndicatorsStyles();
		});

		public static readonly BindableProperty IndicatorsColorProperty = BindableProperty.Create(nameof(IndicatorsColor), typeof(Color), typeof(IndicatorsView), Color.Silver, propertyChanged: (bindable, oldValue, newValue) =>
		{
			(bindable as IndicatorsView)?.ResetIndicatorsStyles();
		});

		public static readonly BindableProperty SelectedIndicatorColorProperty = BindableProperty.Create(nameof(SelectedIndicatorColor), typeof(Color), typeof(IndicatorsView), Color.Gray, propertyChanged: (bindable, oldValue, newValue) =>
		{
			(bindable as IndicatorsView)?.ResetIndicatorsStyles();
		});

		public static readonly BindableProperty IndicatorsSizeProperty = BindableProperty.Create(nameof(IndicatorsSize), typeof(double), typeof(IndicatorsView), 20, propertyChanged: (bindable, oldValue, newValue) =>
		{
			(bindable as IndicatorsView)?.ResetIndicatorsStyles();
		});

		internal static readonly BindableProperty ItemsSourceProperty = BindableProperty.Create(nameof(ItemsSource), typeof(IEnumerable), typeof(IndicatorsView), null, propertyChanged: (bindable, oldValue, newValue) =>
		{
			(bindable as IndicatorsView)?.ResetItemsSource(oldValue as IEnumerable);
		});

		static IndicatorsView()
		{
		}

		public IndicatorsView()
		{
			Spacing = 5;
			Orientation = StackOrientation.Horizontal;
		}

		public int Position
		{
			get => (int)GetValue(PositionProperty);
			set => SetValue(PositionProperty, value);
		}

		public int Count
		{
			get => (int)GetValue(CountProperty);
			set => SetValue(CountProperty, value);
		}

		public int MaxVisibleCount
		{
			get => (int)GetValue(MaxVisibleCountProperty);
			set => SetValue(MaxVisibleCountProperty, value);
		}

		public DataTemplate IndicatorTemplate
		{
			get => GetValue(IndicatorTemplateProperty) as DataTemplate;
			set => SetValue(IndicatorTemplateProperty, value);
		}

		public IndicatorShape IndicatorShape
		{
			get => (IndicatorShape)GetValue(IndicatorShapeProperty);
			set => SetValue(IndicatorShapeProperty, value);
		}

		public Color IndicatorsColor
		{
			get => (Color)GetValue(IndicatorsColorProperty);
			set => SetValue(IndicatorsColorProperty, value);
		}

		public Color SelectedIndicatorColor
		{
			get => (Color)GetValue(SelectedIndicatorColorProperty);
			set => SetValue(SelectedIndicatorColorProperty, value);
		}

		public double IndicatorsSize
		{
			get => (double)GetValue(IndicatorsSizeProperty);
			set => SetValue(IndicatorsSizeProperty, value);
		}

		[EditorBrowsable(EditorBrowsableState.Never)]
		public IEnumerable ItemsSource
		{
			get => GetValue(ItemsSourceProperty) as IEnumerable;
			set => SetValue(ItemsSourceProperty, value);
		}

		public static void SetItemsSourceBy(IndicatorsView indicatorsView, CarouselView carouselView)
		{
			indicatorsView.SetBinding(PositionProperty, new Binding
			{
				Path = nameof(CarouselView.Position),
				Source = carouselView
			});
			indicatorsView.SetBinding(ItemsSourceProperty, new Binding
			{
				Path = nameof(ItemsView.ItemsSource),
				Source = carouselView
			});
		}

		protected virtual void ApplySelectedStyle(View view, int index)
			=> view.BackgroundColor = SelectedIndicatorColor;

		protected virtual void ApplyUnselectedStyle(View view, int index)
			=> view.BackgroundColor = IndicatorsColor;

		private void AddExtraIndicatorsItems()
		{
			var oldCount = Children.Count;
			for (var i = 0; i < Count - oldCount && i < MaxVisibleCount - oldCount; ++i)
			{
				var indicator = IndicatorTemplate?.CreateContent() as View ?? new Frame
				{
					Padding = 0,
					HasShadow = false,
					BorderColor = Color.Transparent,
					WidthRequest = IndicatorsSize,
					HeightRequest = IndicatorsSize,
					CornerRadius = IndicatorShape == IndicatorShape.Circle
						? (float)IndicatorsSize / 2
						: 0
				};
				var itemTapGesture = new TapGestureRecognizer();
				itemTapGesture.Tapped += (tapSender, tapArgs) => Position = IndexOf(tapSender as View);
				indicator.GestureRecognizers.Add(itemTapGesture);
				Children.Add(indicator);
			}
		}

		private void RemoveRedundantIndicatorsItems()
		{
			foreach (var item in Children.Where((v, i) => i >= Count).ToArray())
			{
				Children.Remove(item);
			}
		}

		private int IndexOf(View view) => Children.IndexOf(view);

		private void ApplyStyle(View view)
		{
			try
			{
				view.BatchBegin();
				var index = IndexOf(view);
				if (index == Position)
				{
					ApplySelectedStyle(view, index);
					return;
				}
				ApplyUnselectedStyle(view, index);
			}
			finally
			{
				view.BatchCommit();
			}
		}

		private void ResetIndicatorsStylesNonBatch()
		{
			foreach (var child in Children)
			{
				ApplyStyle(child);
			}
		}

		private void ResetIndicatorsStyles()
		{
			try
			{
				BatchBegin();
				ResetIndicatorsStylesNonBatch();
			}
			finally
			{
				BatchCommit();
			}
		}

		private void ResetIndicatorsCount(int oldValue, int newValue)
		{
			try
			{
				BatchBegin();
				if (oldValue < 0)
				{
					oldValue = 0;
				}

				if (oldValue > newValue)
				{
					RemoveRedundantIndicatorsItems();
					return;
				}

				AddExtraIndicatorsItems();
			}
			finally
			{
				ResetIndicatorsStylesNonBatch();
				BatchCommit();
			}
		}

		private void ResetItemsSource(IEnumerable oldCollection)
		{
			if (oldCollection is INotifyCollectionChanged oldObservableCollection)
			{
				oldObservableCollection.CollectionChanged -= OnObservableCollectionChanged;
			}

			if (ItemsSource is INotifyCollectionChanged observableCollection)
			{
				observableCollection.CollectionChanged += OnObservableCollectionChanged;
			}

			OnObservableCollectionChanged(oldCollection, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
		}

		private void OnObservableCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
		{
			if(sender is ICollection collection)
			{
				Count = collection.Count;
				return;
			}
			var count = 0;
			var enumerator = (sender as IEnumerable)?.GetEnumerator();
			while (enumerator.MoveNext())
			{
				count++;
			}
			Count = count;
		}
	}
}
```

**Comment by samhouts** [comment]
closed by #7465
